tests/suites/test_suite_ssl.function

```diff
@@ -1789,18 +1839,55 @@ void perform_handshake( handshake_test_options* options )
     }
 #endif
 
-    TEST_ASSERT( mbedtls_move_handshake_to_state( &(client.ssl),
-                                                  &(server.ssl),
-                                                  MBEDTLS_SSL_HANDSHAKE_OVER )
-                 == 0 );
+    if( options->use_min_max_version &&
+        options->expected_negotiated_version == MBEDTLS_EXPECT_VERSION_NEGOTIATION_FAIL )
+    {
+        TEST_ASSERT( mbedtls_move_handshake_to_state( &(client.ssl),
+                                                      &(server.ssl),
+                                                      MBEDTLS_SSL_HANDSHAKE_OVER )
+                     == MBEDTLS_ERR_SSL_BAD_HS_PROTOCOL_VERSION );
+
+        /* Connection should have failed by this point, skip to cleanup */
+        goto exit;
+    }
+    else
+    {
+        TEST_ASSERT( mbedtls_move_handshake_to_state( &(client.ssl),
+                                                      &(server.ssl),
+                                                      MBEDTLS_SSL_HANDSHAKE_OVER )
+                     == 0 );
+    }
+
     TEST_ASSERT( client.ssl.state == MBEDTLS_SSL_HANDSHAKE_OVER );
     TEST_ASSERT( server.ssl.state == MBEDTLS_SSL_HANDSHAKE_OVER );
 
+    if( options->use_min_max_version )
+    {
+        /* Check that we agree on the version... */
+        TEST_ASSERT( client.ssl.minor_ver == server.ssl.minor_ver );
+
+        if( client.ssl.minor_ver != options->expected_negotiated_version )
+        {
+            mbedtls_fprintf( stderr,
```

Comment:
It has been a great pain that the TEST_ASSERT macros don't print out more info and I agree that it is nice to be more verbose, I think this would be worth doing in a more generic way. For example we could print out the two compared values in TEST_EQUAL and use that on line 1876.

Reply:
TEST_EQUAL does not place any restriction on the two things passed in, only that they have overloads for == so it really isn't safe to do a printf within TEST_EQUAL as we don't know the types. I will admit I didn't really mean to leave this in, but it has been rather helpful with debugging.

Reply:
Adding a print for `TEST_EQUAL` is the whole reason I introduced it, but we never got around to adding the printf calls (after checking that the type we want to printf is supported everywhere — `long` doesn't cut it on Windows because it's shorter than `size_t`). It's only supposed to be used on integers, so the only uncertainty is signed vs unsigned. It's only debugging output, so even if we end up printing wrapped values, that's better than nothing.

Reply:
@paul-elliott-arm Can you add printing to `TEST_EQUAL` as Gilles suggested? I think casting them to unsigned before printing them would help getting less pain from the compiler and still being safe.

Reply:
@yanesca I fear that it's not that easy: the print would be at the wrong time — normally it's done after the test function returns. On hosted platforms it would just be a display that isn't pretty and may not be on/off at the appropriate level of verbosity. On embedded targets this may break the parsing of test results. So this is out of scope here, and should be done separately, and in a way that works on embedded targets.